examples/tutorial_4_streamproto2.md

```diff
@@ -0,0 +1,106 @@
+In the [previous tutorial](tutorial_3_streamproto.md), we learned how to create a custom streaming protocol.
+
+This tutorial is very similar to the [previous tutorial](tutorial_3_streamproto.md) in that a stream of data will be consumed, however in this tutorial we will learn how to use alternative libp2p streaming semantics.
+
+# Custom protocol in libp2p
+Let's create a `part4.nim`, and import our dependencies:
+```nim
+import bearssl
+import chronos
+import stew/byteutils
+
+import libp2p
+```
+This is similar to the third tutorial. Next, we'll declare our custom protocol.
+```nim
+const TestStreamCodec = "/test/stream-proto/1.0.0"
+
+type TestStreamProto = ref object of LPProtocol
+```
+
+Just as we did in the last tutorial, we've set a [protocol ID](https://docs.libp2p.io/concepts/protocols/#protocol-ids), and created a custom `LPProtocol`.
+
+As in the last tutorial, we are going to handle our client and server parts separately. The server will wait for an instruction from the client, then send a stream of data to the client and finish by closing the connection. The client will listen in a loop until the server has finished and closed the connection. Then the client will close its end of the connection as well.
+
+Let's start with the server part:
+```nim
+# Set the chunk length >= any data being streamed. Our longest chunk of data in
+# either direction is 16 bytes ("please send data").
+const CHUNK_LEN_BYTES = 16
+
+proc new(T: typedesc[TestStreamProto]): T =
+  # every incoming connections will in be handled in this closure
+  var buf = newSeq[byte](CHUNK_LEN_BYTES)
+  proc handle(stream: Connection, proto: string) {.async, gcsafe.} =
+
+    discard await stream.readOnce(addr buf[0], CHUNK_LEN_BYTES)
+    echo "[server] received from client: ", string.fromBytes(buf)
+
+    # send stream of data
+    for i in 0..10:
+      let data = "Part " & $i
+      await stream.write(data)
+
+    # We must close the connections ourselves when we're done with it
+    await stream.close()
+
+  return T(codecs: @[TestStreamCodec], handler: handle)
+```
+Instead of using `readLp` to read data from the connection (as in the previous tutorial), we are going to use `readOnce`, which allows us to read bytes from the buffer or wait for bytes to be received.
+
+We can now create our client part:
+```nim
+proc streamData(p: TestStreamProto, stream: Connection) {.async.} =
+  let data = "please send data"
+  await stream.write(data)
+
+  # Read loop
+  while stream.atEof.not:
```

Comment:
```suggestion
  while not stream.atEof:
```
Maybe it's just me, but I find this syntax a bit better, especially to people whom may not be introduced to nim

Reply:
Updated.